### PR TITLE
Feature/refactor controllers

### PR DESCRIPTION
**Pull request by ifalatik**
@dylanturn I'm in the process of refactoring the controllers.

I'm currently done with the firewall alias & filter controllers. I just created this PR so you can already look at it, and give some feedback.

### CODE REVIEW

**Comment by dylanturn** [comment]
Fantastic! I'll set aside some time today to review.

By the way, I've updated the readme to include an invite link to the Turnbros discord server. I figured it might be easier to coordinate over Discord than via Github PR comment. Up to you though, I'm happy with whatever you prefer. 

Thanks again!

**Comment by ifalatik** [comment]
Awesome, I'll join right away!

File: opnsense_api/interfaces/loopback_controller.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ class LoopbackInterface(OPNsenseItem):
 class LoopbackController(OPNsenseItemController[LoopbackInterface]):
 
     def __init__(self, device):
-        super().__init__(device, "interfaces", "loopback_settings")
+        super().__init__(device, "interface", "loopback_settings")
```
Comment (dylanturn) [inline opnsense_api/interfaces/loopback_controller.py]: FYI, it was `interfaces` because that's the name of the module in the [API](https://docs.opnsense.org/development/api/core/interfaces.html). The first argument is module, second is controller. Those two values are used when formatting the API request.

File: opnsense_api/interfaces/vip_controller.py [hunk 2]
```diff
@@ -23,15 +23,15 @@ class VIPInterface(OPNsenseItem):
 class VIPController(OPNsenseItemController[VIPInterface]):
 
     def __init__(self, device):
-        super().__init__(device, "interfaces", "vip_settings")
+        super().__init__(device, "interface", "vip_settings")
 
     def _parse_api_response(self, api_response) -> VIPInterface:
         return VIPInterface(
         uuid=api_response["uuid"],
-        description=api_response["descr"],
```
Comment (dylanturn) [inline opnsense_api/interfaces/vip_controller.py:31]: `_parse_api_response` is parsing the item JSON object sent in the devices response, which doesn't include `description`. For whatever reason it seems for that item the description key was abbreviated to `descr`

File: opnsense_api/util/__init__.py [hunk 3]
```diff
@@ -6,13 +6,6 @@ class Constants(object):
     HTTP_SUCCESS = [200, 201, 202, 203, 204, 205, 206, 207]
 
 
-class AliasType(Enum):
-    HOST = "host"
-    NETWORK = "network"
-    PORT = "port"
-    URL = "url"
```
Comment (dylanturn) [inline opnsense_api/util/__init__.py:13]: Was there a reason `AliasType` was removed entirely?

File: opnsense_api/firewall/alias_controller.py [hunk 4]
```diff
@@ -1,182 +1,110 @@
...
+    "BGP ASN": "asn",
+    "Dynamic IPv6 Host": "dynipv6host",
+    "Internal (automatic)": "internal",
+    "External (advanced)": "external"
+}
+
+
+class Alias(OPNSenseItem):
+    name: constr(min_length=1, max_length=32, strip_whitespace=True, regex=r"^[a-zA-Z0-9_]*$")
+    type: constr(to_lower=True)
```
Comment (dylanturn) [inline opnsense_api/firewall/alias_controller.py]: Is there a reason we removed it as an enum? I had made it an enum because the number of possible types is fixed and using an enum would make it easier for users of the library to know what alias types were allowed.

File: opnsense_api/util/__init__.py (hunk 3 above)
Comment (dylanturn) [reply opnsense_api/util/__init__.py:13]: I eventually discovered `LONG_NAME_TO_TYPE_DICT`, but I'm not really sure that's an improvement.

File: opnsense_api/util/controller.py [hunk 5]
```diff
@@ -1,24 +1,24 @@
...
         # Simplest url path for a request
         # e.g. api/unbound/settings/searchHostOverride
         base_request = f"{self._module}/{self._controller}/{command}"
 
         # Add in the UUID for a specific resource
         # e.g. api/unbound/settings/getHostOverride/2ce9672e-43a5-4462-9cf1-084964970862
-        if uuid is not None:
-            base_request = f"{self._module}/{self._controller}/{command}/{uuid}"
+        if resource_identifier is not None:
+            base_request = f"{base_request}/{resource_identifier}"
```
Comment (dylanturn) [inline opnsense_api/util/controller.py]: What was the motivation behind this? Is this change related to the edge cases you mentioned?

**Review by dylanturn — COMMENTED** [review]
I'm not quite done reviewing it, There a lot that's changed so it's a challenge for me to parse through and make sure I understand it all. Almost feels like a totally different code base.

File: opnsense_api/interfaces/loopback_controller.py (hunk 1 above)
Comment (ifalatik) [reply opnsense_api/interfaces/loopback_controller.py]: Yeah, i have no idea why my IDE changed that. Definitely wasn't a conscious decision on my part 😂. I'll change it back once I get to these controllers.

File: opnsense_api/util/controller.py (hunk 5 above)
Comment (ifalatik) [reply opnsense_api/util/controller.py]: Well for one I wanted to not recreate what we already have in base_request. But I guess your question is related to uuid -> resource_identifier. I think that came from me mistyping a uuid during API testing and thinking: "Oh, it seems like it isn't always the uuid. I'll look over everything again, and can revert it if it is indeed the uuid everywhere, if you wish.

**Comment by ifalatik** [comment]
If I were you I'd start with the OPNSenseItemController, then OPNSenseApplicableItemController.
Afterwards AliasController then FilterController. That should help understanding (most of) the changes.

File: opnsense_api/firewall/alias_controller.py (hunk 4 above)
Comment (ifalatik) [reply opnsense_api/firewall/alias_controller.py]: Yes. Whenever you call searchItem for these, it returns the type as one of the "long names". This allows for easy mapping to the names used during adding and setting. Also, we can use the values to validate any new alias the user creates.

**Comment by ifalatik** [comment]
Also there's likely going to be minor changes in all of the stated classes, as I refactor more controllers. I'm trying to reduce code as much as possible here, but don't really want to end up with too many workarounds for generic solutions like I had to use in FilterController:
Whenever you call list() it returns FilterRuleBase objects, for getting, setting, deletion, etc. it uses FilterRule objects. But look at this once you've looked at and understood the other classes, so you don't get confused as to why I did it the way I did.

**Comment by ifalatik** [comment]
For an example of how pretty this can look, take a look at the "new and improved" category controller

File: opnsense_api/util/controller.py (hunk 5 above)
Comment (dylanturn) [reply opnsense_api/util/controller.py]: lets revert that, `resource_identifier` is going to be confusing in the long run.

(I actually meant to leave this comment a couple weeks ago but I didn't notice it was "pending")

File: opnsense_api/interfaces/vip_controller.py (hunk 2 above)
Comment (dylanturn) [reply opnsense_api/interfaces/vip_controller.py:31]: Did we address this? I don't remember now...

**Review by dylanturn — APPROVED** [review]
mega effort!
